src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,974 @@
+            var declarator = expression?.Parent?.Parent;
+
+            localDeclaration = null;
+
+            if (!syntaxFacts.IsVariableDeclarator(declarator))
+            {
+                varDeclName = null;
+                return false;
+            }
+
+            localDeclaration = GetLocalDeclarationFromDeclarator(declarator);
+            if (localDeclaration is null)
+            {
+                varDeclName = null;
+                return false;
+            }
+
+            // TODO: handle in the future
+            if (syntaxFacts.GetVariablesOfLocalDeclarationStatement(localDeclaration).Count > 1)
+            {
+                varDeclName = null;
+                localDeclaration = null;
+                return false;
+            }
+
+            varDeclName = syntaxFacts.GetIdentifierOfVariableDeclarator(declarator).ValueText;
+            return true;
+        }
+
+        /// <summary>
+        /// Locates all the call sites of the method that introduced the parameter
+        /// </summary>
+        protected static async Task<Dictionary<Document, List<SyntaxNode>>> FindCallSitesAsync(
+            Document document, IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var methodCallSites = new Dictionary<Document, List<SyntaxNode>>();
+            var progress = new StreamingProgressCollector();
+            await SymbolFinder.FindReferencesAsync(
+                methodSymbol, document.Project.Solution, progress,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+            var referencedSymbols = progress.GetReferencedSymbols();
+
+            // Ordering by descending to sort invocations by starting span to account for nested invocations
+            var referencedLocations = referencedSymbols.SelectMany(referencedSymbol => referencedSymbol.Locations)
+                .Distinct().Where(reference => !reference.IsImplicit)
+                .OrderByDescending(reference => reference.Location.SourceSpan.Start);
+
+            // Adding the original document to ensure that it will be seen again when processing the call sites
+            // in order to update the original expression and containing method.
+            methodCallSites.Add(document, new List<SyntaxNode>());
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            foreach (var refLocation in referencedLocations)
+            {
+                // Does not support cross-language references currently
+                if (refLocation.Document.Project.Language == document.Project.Language)
+                {
+                    var reference = refLocation.Location.FindNode(cancellationToken).GetRequiredParent();
+                    var originalReference = reference;
+                    if (reference is not (TObjectCreationExpressionSyntax or TInvocationExpressionSyntax))
```

Comment:
interesting.  do you have any tests where teh *callee* does something like `this.Foo(x, y z)`.  I would have expecte the reference to point at `Foo` and GetRequiredParent to return `this.Foo`, not the invocation.  Can you verify (or add) test here?

Reply:
I have a test for that, it's called TestExpressionWithSingleMethodCallAndAccessorsTrampoline. It does as you say, and then the parent of this.Foo is the invocation

Reply:
What about a.b.Foo?

Reply:
yep, TestExpressionWithSingleMethodCallMultipleAccessorsTrampoline